hydra-test-utils/src/Test/Network/Ports.hs

```diff
@@ -40,61 +21,35 @@ import System.Random.Shuffle (
 -- port returned by 'getRandomPort' before this process does.
 --
 -- Do not use this unless you have no other option.
-getRandomPort :: IO PortNumber
-getRandomPort = do
+getRandomPort :: TVar IO [Int] -> IO PortNumber
+getRandomPort tvar = do
   (port, sock) <- openFreePort
   liftIO $ close' sock
-  return $ fromIntegral port
+  mport <- atomically $ recordFreePort tvar port
+  case mport of
+    Nothing -> threadDelay 1 >> getRandomPort tvar
+    Just p -> return $ fromIntegral p
 
 -- | Find a free TCPv4 port and pass it to the given 'action'.
 --
--- Should be used only for testing, see 'getRandomPort' for limitations.
-withFreePort :: (PortNumber -> IO ()) -> IO ()
-withFreePort action = getRandomPort >>= action
-
--- | Checks whether @connect()@ to a given TCPv4 `SockAddr` succeeds or
--- returns `eCONNREFUSED`.
---
--- Rethrows connection exceptions in all other cases (e.g. when the host
--- is unroutable).
---
--- Code courtesy of nh2: https://stackoverflow.com/a/57022572
-isPortOpen :: SockAddr -> IO Bool
-isPortOpen sockAddr = do
-  bracket (socket AF_INET Stream 6 {- TCP -}) close' $ \sock -> do
-    res <- try $ connect sock sockAddr
-    case res of
-      Right () -> return True
-      Left e
-        | (Errno <$> ioe_errno e) == Just eCONNREFUSED -> pure False
-        | "WSAECONNREFUSED" `isInfixOf` show e -> pure False
-        | otherwise -> throwIO e
+-- NOTE: Should be used only for testing, see 'getRandomPort' for limitations.
+withFreePort :: TVar IO [Int] -> (PortNumber -> IO a) -> IO a
+withFreePort tvar action = getRandomPort tvar >>= action
 
--- | Creates a `SockAttr` from host IP and port number.
+-- | Find the specified number of free ports.
 --
--- Example:
--- > simpleSockAddr (127,0,0,1) 8000
-simpleSockAddr :: (Word8, Word8, Word8, Word8) -> PortNumber -> SockAddr
-simpleSockAddr addr port = SockAddrInet (fromIntegral port) (tupleToHostAddress addr)
-
--- | Get a list of random TCPv4 ports that currently do not have any servers
--- listening on them. It may return less than the requested number of ports.
---
--- Note that this method of allocating ports is subject to race
--- conditions. Production code should use better methods such as passing a
--- listening socket to the child process.
-randomUnusedTCPPorts :: Int -> IO [Int]
-randomUnusedTCPPorts count = do
```

Comment:
Isn't this the source of all race conditions? The documentation of `openFreePort` and especially on [getFreePort](https://hackage.haskell.org/package/port-utils-0.2.1.0/docs/Network-Socket-Free.html#v:getFreePort) (a factored out helper from `warp` we should use`) as well as our code states that the race condition is only when other processes do bind that port between `getFreePort` and binding to it again.

HOWEVER, this code here seemingly is still used (as you updated it), does also tests ports as open .. on a shuffled, self-made list! This basically circumvents the mechanism used in `openFreePort`/`getFreePort`, which asks the OS to give a free port. The OS does not know of the list we created in this function!